python/tests/utils/power_spectrum.py

```diff
@@ -0,0 +1,134 @@
+
+
+def se():
+    return se_calculator().compute(SystemForTests())
+
+
+def ps():
+    return PowerSpectrum(se_calculator()).compute(SystemForTests())
+
+
+@pytest.mark.parametrize("calculator", [se_calculator(), lode_calculator()])
+def test_power_spectrum(calculator) -> None:
+    """Test that power spectrum works and that the shape is correct."""
+    ps_python = PowerSpectrum(calculator).compute(SystemForTests())
+    ps_python = ps_python.keys_to_samples(["species_center"])
+
+    # Test the number of properties is correct
+    n_props_actual = len(ps_python.block().properties)
+
+    n_props_expected = (
+        N_SPECIES**2 * HYPERS["max_radial"] ** 2 * (HYPERS["max_angular"] + 1)
+    )
+
+    assert n_props_actual == n_props_expected
+
+
+def test_error_max_angular():
+    """Test error raise if max_angular are different."""
+    hypers_2 = HYPERS.copy()
+    hypers_2.update(max_radial=3, max_angular=1)
+
+    se_calculator2 = rascaline.SphericalExpansion(**hypers_2)
+
+    msg = "'max_angular' of both calculators must be the same!"
+    with pytest.raises(ValueError, match=msg):
+        PowerSpectrum(se_calculator(), se_calculator2)
+
+    calculator = rascaline.SoapPowerSpectrum(**HYPERS)
+    with pytest.raises(ValueError, match="are supported for calculator_1"):
+        PowerSpectrum(calculator)
+
+
+def test_wrong_calculator_1():
+    """Test error raise for wrong calculator_1."""
+
+    calculator = rascaline.SoapPowerSpectrum(**HYPERS)
+    with pytest.raises(ValueError, match="are supported for calculator_1"):
+        PowerSpectrum(calculator)
+
+
+def test_wrong_calculator_2():
+    """Test error raise for wrong calculator_2."""
+
+    calculator = rascaline.SoapPowerSpectrum(**HYPERS)
+    with pytest.raises(ValueError, match="are supported for calculator_2"):
+        PowerSpectrum(se_calculator(), calculator)
+
+
+def test_power_spectrum_different_hypers() -> None:
+    """Test that power spectrum works with different spherical expansions."""
+
+    hypers_2 = HYPERS.copy()
+    hypers_2.update(max_radial=3, max_angular=4)
+
+    se_calculator2 = rascaline.SphericalExpansion(**hypers_2)
+
+    PowerSpectrum(se_calculator(), se_calculator2).compute(SystemForTests())
+
+
+def test_power_spectrum_rust() -> None:
+    """Test that the dot kernels of the rust and python version are the same."""
+
+    ps_python = ps()
+    ps_python = ps_python.keys_to_samples(["species_center"])
+    kernel_python = np.dot(ps_python[0].values, ps_python[0].values.T)
+
+    ps_rust = rascaline.SoapPowerSpectrum(**HYPERS).compute(SystemForTests())
+    ps_rust = ps_rust.keys_to_samples(["species_center"])
+    ps_rust = ps_rust.keys_to_properties(["species_neighbor_1", "species_neighbor_2"])
+    kernel_rust = np.dot(ps_rust[0].values, ps_rust[0].values.T)
+    assert_allclose(kernel_python, kernel_rust)
+
+
+def test_power_spectrum_gradients() -> None:
+    """Test that gradients are correct using finite differences."""
+    # TODO
+    pass
```

Comment:
Do you want to finish this before merging?

Reply:
There is a Rust framework for finite differences here: https://github.com/Luthaf/rascaline/blob/9c2eb78af9a1564765b8381ac4acdfdaa079e21f/rascaline/src/calculators/tests_utils.rs#L251